lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+      auto CallerArchetype = dyn_cast<SubstitutableType>(
+          OrigGenericEnv->mapTypeIntoContext(Req.getFirstType())
+              ->getCanonicalType());
+      // Remember that a given generic parameter is mapped
+      // to a concrete type.
+      CallerArchetypeToInterfaceMap.addSubstitution(
+          CallerArchetype, Req.getSecondType()->getCanonicalType());
+
+      // Remember how the original interace type is mapped to a concrete type.
+      ClonerArchetypeToConcreteMap.addSubstitution(
+          CallerArchetype, Req.getSecondType()->getCanonicalType());
+
+      // Remember how the original interace type is mapped to a concrete type.
+      CallerArchetypeToConcreteMap.addSubstitution(
+          CallerArchetype, Req.getSecondType()->getCanonicalType());
+    }
+  }
+
+  std::tie(SpecializedGenericEnv, SpecializedGenericSig) =
+      getSignatureWithRequirements(OrigGenericSig, OrigGenericEnv,
+                                   Requirements, M);
+
+  for (auto Req : SpecializedGenericSig->getRequirements()) {
+    // Remember how the original contextual type is represented in
+    // the specialized function.
+    if (Req.getKind() == RequirementKind::Layout) {
+      auto CallerArchetype = dyn_cast<SubstitutableType>(
+          OrigGenericEnv->mapTypeIntoContext(Req.getFirstType())
+              ->getCanonicalType());
+
+      ClonerArchetypeToConcreteMap.addSubstitution(
+          CallerArchetype,
+          SpecializedGenericEnv->mapTypeIntoContext(Req.getFirstType()));
+
+      CallerArchetypeToInterfaceMap.addSubstitution(
+          CallerArchetype,
+          SpecializedGenericEnv->mapTypeOutOfContext(Req.getFirstType()));
+    }
+  }
+
+  // If this is a partial specialization, some of the generic type parameters
+  // are not substituted, so add the missing substitutions which basically map
+  // the caller archetypes to their interface types or the corresponding
+  // archetypes in the specialized function.
+  for (auto GP : SpecializedGenericSig->getGenericParams()) {
+    auto CallerArchetype = dyn_cast<SubstitutableType>(
+        OrigGenericEnv->mapTypeIntoContext(GP)->getCanonicalType());
+    if (CallerArchetypeToInterfaceMap.lookupSubstitution(CallerArchetype))
+      continue;
+    CallerArchetypeToInterfaceMap.addSubstitution(CallerArchetype,
+                                                  GP->getCanonicalType());
+    if (ClonerArchetypeToConcreteMap.lookupSubstitution(CallerArchetype))
+      continue;
+    auto SpecializedArchetype =
+        SpecializedGenericEnv->mapTypeIntoContext(GP);
+    ClonerArchetypeToConcreteMap.addSubstitution(CallerArchetype,
+                                                 SpecializedArchetype);
+  }
+
+  // Substitute into forwarding substitutions to get the cloner substitutions
+  // and substitutions to be used by the caller for calling the specialized
+  // function.
+  SmallVector<Substitution, 4> ClonerSubsList;
```

Comment:
Instead, construct these arrays by calling GenericSignature::getSubstitutions() or GenericEnvironment::getSubstitutions(). Otherwise you're relying on them having entries in the same order as the forwarding substitutions.